### ISSUE TITLE
Stop Fail messages from overriding the existing test message

### ISSUE THREAD

**Issue by dc95ie** [robotframework/robotframework#4164]
My test cases print important information in the Test Message column of the Robot report, regardless of whether the test passes or fails. We use Robot for automated hardware production testing and have an end-user that reviews these reports. It is important for them to have quick access to this information, rather than trawling through a log to find what they are after. For example, we use the Test Message column to display important calibration statistics, system serials and hyperlinks to graphs plotted during the Robot test run.

If a test fails, all this information is wiped clean and the test message is overridden in the report with the Fail message. Can we consider adding a 'prepend' option (similar to the 'append' option of the Set Test Message keyword) that will add the Fail message to the start of the existing test message, rather than wiping it clean?

**Comment by pekkaklarck** [comment]
We could easily add configuration options controlling the failure message to individual keywords like `Fail`, but it's impossible to add such configuration to all keywords. I guess we could consider changing the overall logic so that possible earlier message is preserved as part of the final message. For example, if a test would have a message like "Example" and then it would fail with message "Failure", the final message could be like this:

```
Failure

Original message was:
Message
```

We've used similar approach, for example, when a test teardown fails and the test has already failed. Adding this functionality shouldn't be overly complicated, but there could be interesting corner cases to take into account with teardown failures, skipping, etc.

I personally don't have time to look at this, but if you @dc95ie or someone else is interested to investigate how to add this I'm happy to help.

**Comment by dc95ie** [comment]
Yes, it would be good to preserve the original test message in the way you have described.

My current workaround is to save the existing test message to a variable before calling the Fail keyword, and then recalling this variable in the Test Teardown and manually adding it to the end of the Fail message. This works OK, but is obviously cumbersome. 

I will work on adding in this feature as time permits.

**Comment by nick-grout1** [comment]
Hi @pekkaklarck, @dc95ie - I was wondering what the status of this is. I would love to implement this @dc95ie if you're too busy to! 

Just let me know if you'd like me to take a stab at it and I'll be happy to take a look.

**Comment by ericbjones** [comment]
I don't think the default behavior of a test failing naturally and then overwriting output is wrong.  Especially since the default behavior of passing does not set a test message.  IMO this issue applies most or is most concerning to a specific use cases where people are using `Set Test Message` with the `append=true` parameter.  `Set Test Message` allows you to modify the message without passing *yet* and keep going and building up a test message.  We should at least have an option int the `FAIL` keyword to do the same `append=True` behavior.  Then it would be easy for test writers to avoid failures happening naturally by checking conditions themselves or using the `And Return Status` or similar types of `Run Keyword` commands (Soon to be much better with Try Except in RF5) and then manually `FAIL` with `append=True`.  That's all we need to make this acceptable in my mind.  I don't think everyone needs this in every use case.  Just need to ability to use `FAIL` keyword in the same way as `Set Test Message` ... and while we are at it `PASS` should also have the `append=True` option.

**Comment by ericbjones** [comment]
@pekkaklarck anything in RF5 already planned that changes test report outputs that might improve this? or does this one still need an owner?

**Comment by dc95ie** [comment]
@ericbjones, I can take a look at implementing this over the next couple of days. A simple `append=True` option for the FAIL message would be perfectly fine for my application. 

FYI, as a workaround, I am currently saving test messages to a global variable during the run sequence. At Teardown, I copy this variable into the test message using `Set Test Message append=True`. This will copy the Test Message contents that have accumulated during the body of the test below the Fail message.
